app/src/main/java/org/lineageos/glimpse/ViewActivity.kt

```diff
@@ -54,118 +306,363 @@ class ViewActivity : AppCompatActivity() {
+        )
     }
 
     /**
      * Handle a [MediaStore.ACTION_REVIEW] / [MediaStore.ACTION_REVIEW_SECURE] intent
      * (view a media together with medias from the same bucket ID).
      * If uri parsing from [MediaStore] fails, fallback to [handleView].
-     * Must be executed on [ioScope].
+     * Must not be executed on main thread.
      * @param intent The received intent
      * @param secure Whether we should review this media in a secure manner
+     * @return true if the intent has been handled, false otherwise
      */
-    private fun handleReview(intent: Intent, secure: Boolean = false) {
-        intent.data?.let { getMediaStoreMedia(it) }?.also {
-            val additionalMedias = intent.clipData?.asArray()?.mapNotNull {
-                getMediaStoreMedia(it.uri)
-            }
-            runOnUiThread {
-                supportFragmentManager
-                    .beginTransaction()
-                    .replace(
-                        R.id.navHostFragment, MediaViewerFragment.newInstance(
-                            it,
-                            it.bucketId.takeUnless { secure },
-                            additionalMedias,
-                            null,
-                            secure
-                        )
-                    )
-                    .commit()
-            }
-        } ?: handleView(intent, secure)
+    private fun handleReview(intent: Intent, secure: Boolean = false) = intent.data?.let {
+        getMediaStoreMedia(it)
+    }?.let { media ->
+        val additionalMedias = intent.clipData?.asArray()?.mapNotNull {
+            getMediaStoreMedia(it.uri)
+        }
+
+        updateArguments(
+            media = media,
+            albumId = intent.extras?.getInt(KEY_ALBUM_ID, -1)?.takeUnless {
+                it == -1
+            } ?: media.bucketId.takeUnless { secure },
+            additionalMedias = additionalMedias?.toTypedArray()?.takeIf { it.isNotEmpty() },
+            secure = secure,
+        )
+    } ?: handleView(intent, secure)
+
+    /**
+     * Update all media arguments and make sure stuff is set properly.
+     * @param media The first media to show
+     * @param albumId Album ID, if null [additionalMedias] will be used
+     * @param additionalMedias The additional medias to show alongside [media]
+     * @param mediaUri The [MediaUri] to show
+     * @param secure Whether this should be considered a secure session
+     */
+    private fun updateArguments(
+        media: Media? = null,
+        albumId: Int? = null,
+        additionalMedias: Array<Media>? = null,
+        mediaUri: MediaUri? = null,
+        secure: Boolean = false,
+    ): Boolean {
+        this.media = media
+        this.albumId = albumId
+        this.additionalMedias = additionalMedias
+        this.mediaUri = mediaUri
+        this.secure = secure
+
+        // Update UI while at it
+        runOnUiThread {
+            updateUI()
+        }
+
+        return true
```

Comment:
always true?

Reply:
Moved return value to each intent handler method